nexus/db-queries/src/db/datastore/target_release.rs

```diff
@@ -0,0 +1,265 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods to get/set the current [`TargetRelease`].
+
+use super::DataStore;
+use crate::authz;
+use crate::context::OpContext;
+use crate::db::error::{public_error_from_diesel, ErrorHandler};
+use crate::db::model::{SemverVersion, TargetRelease, TargetReleaseSource};
+use crate::db::schema::target_release;
+use crate::db::schema::target_release::dsl;
+use async_bb8_diesel::AsyncRunQueryDsl as _;
+use diesel::insert_into;
+use diesel::prelude::*;
+use nexus_types::external_api::views;
+use omicron_common::api::external::{CreateResult, Error, LookupResult};
+
+impl DataStore {
+    /// Fetch the current target release, i.e., the row with the largest
+    /// generation number.
+    pub async fn target_release_get_current(
+        &self,
+        opctx: &OpContext,
+    ) -> LookupResult<TargetRelease> {
+        opctx
+            .authorize(authz::Action::Read, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        // Fetch the row in the `target_release` table with the largest
+        // generation number. The subquery accesses the same table, so we
+        // have to make an alias to not confuse diesel.
+        let target_release_2 =
+            diesel::alias!(target_release as target_release_2);
+        dsl::target_release
+            .filter(dsl::generation.nullable().eq_any(target_release_2.select(
+                diesel::dsl::max(target_release_2.field(dsl::generation)),
+            )))
+            .select(TargetRelease::as_select())
+            .first_async(&*conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
```

Comment:
I'm not sure how much this matters but I would think you could do something like we do in `blueprint_current_target_only()`:

```rust
dsl::target_release
    .order_by(dsl::generation.desc())
    .first_async()
```

This would avoid an extra subquery (and the need to use an alias).

---

Separately: what does this do when the table is empty?  (What do we want it to do?  I see that `blueprint_current_target_only()` explicitly allows this to be optional but then converts `None` to a specific error.  I'm not sure if we want to make this a similar sort of 500 or return a (fake?) initial target release of "unspecified".)

Reply:
Thanks, that's much simpler without the subquery! Both issues should be addressed in e7181dbbe.